internal/testutil/runmain.go

```diff
@@ -59,46 +63,55 @@ func (r *Runner) Cleanup() {
 	}
 }
 
-// Run runs the built binary with the given environment.
-// After f returns, the running process is shut down.
-func (r *Runner) Run(env map[string]string, f func()) {
+// Stdout returns the stdout from the most recent Run()
+func (r *Runner) Stdout() []byte {
+	return r.stdout.Bytes()
+}
+
+// Stderr returns the stderr from the most recent Run()
+func (r *Runner) Stderr() []byte {
+	return r.stderr.Bytes()
+}
+
+// Run executes runs the built binary until terminated or timeout has
+// been reached, and indicates successful execution on return.
+func (r *Runner) Run(env map[string]string, timeout time.Duration) error {
 	if !r.Built() {
-		r.t.Error("Tried to run when binary not built.")
-		return
+		return fmt.Errorf("tried to run when binary not built")
 	}
+	r.stderr.Reset()
+	r.stdout.Reset()
 	environ := os.Environ()
 	for k, v := range env {
 		environ = append(environ, k+"="+v)
 	}
 
-	cmd := exec.Command(r.bin)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, r.bin)
 	cmd.Env = environ
 
-	if err := cmd.Start(); err != nil {
-		r.t.Error(err)
-		return
+	stdErr, err := cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("could not get stderr pipe")
+	}
+	stdOut, err := cmd.StdoutPipe()
```

Comment:
cmd.Stdout = r.stdout
cmd.Stderr = r.stderr

No need to mess around with pipes.